client_test.go

```diff
@@ -644,39 +644,83 @@ func Test_Client(t *testing.T) {
 	t.Run("StopAndCancel", func(t *testing.T) {
 		t.Parallel()
 
-		client, _ := setup(t)
-		jobStartedChan := make(chan int64)
-		jobDoneChan := make(chan struct{})
-
-		type JobArgs struct {
-			JobArgsReflectKind[JobArgs]
+		type testBundle struct {
+			jobDoneChan    chan struct{}
+			jobStartedChan chan int64
 		}
 
-		AddWorker(client.config.Workers, WorkFunc(func(ctx context.Context, job *Job[JobArgs]) error {
-			jobStartedChan <- job.ID
-			<-ctx.Done()
-			require.ErrorIs(t, context.Cause(ctx), rivercommon.ErrShutdown)
-			close(jobDoneChan)
-			return nil
-		}))
+		setupStopAndCancel := func(t *testing.T) (*Client[pgx.Tx], *testBundle) {
+			t.Helper()
 
-		startClient(ctx, t, client)
+			client, _ := setup(t)
+			jobStartedChan := make(chan int64)
+			jobDoneChan := make(chan struct{})
 
-		insertRes, err := client.Insert(ctx, &JobArgs{}, nil)
-		require.NoError(t, err)
+			type JobArgs struct {
+				JobArgsReflectKind[JobArgs]
+			}
 
-		startedJobID := riverinternaltest.WaitOrTimeout(t, jobStartedChan)
-		require.Equal(t, insertRes.Job.ID, startedJobID)
+			AddWorker(client.config.Workers, WorkFunc(func(ctx context.Context, job *Job[JobArgs]) error {
+				jobStartedChan <- job.ID
+				<-ctx.Done()
+				require.ErrorIs(t, context.Cause(ctx), rivercommon.ErrShutdown)
+				close(jobDoneChan)
+				return nil
+			}))
 
-		select {
-		case <-client.Stopped():
-			t.Fatal("expected client to not be stopped yet")
-		default:
+			startClient(ctx, t, client)
+
+			insertRes, err := client.Insert(ctx, &JobArgs{}, nil)
+			require.NoError(t, err)
+
+			startedJobID := riverinternaltest.WaitOrTimeout(t, jobStartedChan)
+			require.Equal(t, insertRes.Job.ID, startedJobID)
+
+			select {
+			case <-client.Stopped():
+				t.Fatal("expected client to not be stopped yet")
+			default:
+			}
+
+			return client, &testBundle{
+				jobDoneChan:    jobDoneChan,
+				jobStartedChan: jobStartedChan,
+			}
 		}
 
-		require.NoError(t, client.StopAndCancel(ctx))
+		t.Run("OnItsOwn", func(t *testing.T) {
+			t.Parallel()
+
+			client, _ := setupStopAndCancel(t)
+
+			require.NoError(t, client.StopAndCancel(ctx))
+			riverinternaltest.WaitOrTimeout(t, client.Stopped())
+		})
+
+		t.Run("AfterStop", func(t *testing.T) {
+			t.Parallel()
+
+			client, bundle := setupStopAndCancel(t)
+
+			go func() {
+				require.NoError(t, client.Stop(ctx))
+			}()
+
+			select {
+			case <-client.Stopped():
+				t.Fatal("expected client to not be stopped yet")
+			case <-time.After(500 * time.Millisecond):
```

Comment:
Yeah, I'd like to avoid these as well. Do you have an idea of how else to test this bit reliably and quickly? Specifically, making sure the client is _not_ yet stopped and where I need to call `Stop()` in another goroutine.

Reply:
> With this sleep statement, the test case requires a minimum of 500 ms to run. And latency aside, it also opens the door to intermittency problems -- is 500 ms a good number for both CI and local? If so, what about 250 ms? If that's okay, how about 125 ms? We're kind of just picking an arbitrary number and hoping it's long enough.

On this point, the intermittency is a little different with these cases where we're trying to make sure that something _hasn't_ happened than the cases where we're trying to wait to ensure that something _does_ happen. The reason is that any intermittency will at least surface itself as a flaky test that sometimes fails on something that _should never happen_, so that alone indicates something broken.

Of course it would be far preferable to be able to test this condition in such a way that is ~instant and 100% deterministic—I'm just not sure how to make that work cleanly with the primitives we've exposed here. For example, how can we make sure that we're waiting until the `Stop()` has been initiated? Would you want to expand the surface area of `BaseStartStop` to expose this so it's available in tests? I think doing so would require some redesign of those internals given the way its mutex locks are currently structured.

Reply:
> For example, how can we make sure that we're waiting until the `Stop()` has been initiated? Would you want to expand the surface area of `BaseStartStop` to expose this so it's available in tests? I think doing so would require some redesign of those internals given the way its mutex locks are currently structured.

I'm definitely not against this. The start/stop should only present the most minimal viable interface publicly, but we can add a second interface that reveals a few more internals for test-related purposes. I can take looking into this separately.